Merge Payment Requests into 2.0

GSadee opened the pull request:
| Q               | A
|-----------------|-----
| Branch?         | 2.0
| Bug fix?        | no
| New feature?    |yes
| BC breaks?      | no
| Deprecations?   | no<!-- don't forget to update the UPGRADE-*.md file -->
| Related tickets | 
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
 - Features, removing deprecations and BC breaks must be submitted against the 2.0 branch
 - Make sure that the correct base branch is set

 To be sure you are not breaking any Backward Compatibilities, check the documentation:
 https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

GSadee committed:
Add CypherInterface to composer require checker whitelist

GSadee committed:
Move GatewayConfig validation configs to PaymentBundle

GSadee committed:
Fix moving AsGatewayConfigurationType attribute to PaymentBundle

GSadee committed:
[Payment] Minor fixes after CR

GSadee committed:
[Payment] Extract generator for gateway name

GSadee committed:
[Payment] Refactor generating gateway config validation groups

TheMilek committed:
Move GatewayConfig to Payment component (#16695)

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | partially
https://github.com/Sylius/Sylius/issues/16237
| License         | MIT

TODO:

- [x] Move validation configs to the PaymentBundle

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
- Features, removing deprecations and BC breaks must be submitted
against the 2.0 branch
 - Make sure that the correct base branch is set

To be sure you are not breaking any Backward Compatibilities, check the
documentation:

https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

Prometee committed:
Add deprecation notice mention by @diimpp

Prometee committed:
Move the SyliusPayumBundle Order pay controller action to a dedicated ShopBundle controller action

# Conflicts:
#	UPGRADE-2.0.md

# Conflicts:
#	src/Sylius/Bundle/PayumBundle/Controller/PayumController.php
#	src/Sylius/Bundle/PayumBundle/Resources/config/services/controller.xml

Prometee committed:
Replace else by a return

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

Prometee committed:
Format new line

Co-authored-by: Dmitri Perunov <diimpp@gmail.com>

Prometee committed:
Explicit assert

Co-authored-by: Dmitri Perunov <diimpp@gmail.com>

Prometee committed:
Format new line

Co-authored-by: Dmitri Perunov <diimpp@gmail.com>

Prometee committed:
Simplify gateway check

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

Prometee committed:
Remove default response provider and introduce LogicException instead

Prometee committed:
Fix typo and shorten return

Prometee committed:
Refactor and enhance thrown message

Prometee committed:
Add PaymentRequest HttpResponseProviders

Prometee committed:
Add PaymentToPayResolver to get the last payment via a service

Prometee committed:
Add payment request http response provider

Prometee committed:
Add order tokenValue to the after pay route

Prometee committed:
Specify type

Prometee committed:
Move payum controller body to a dedicated service

Prometee committed:
Create new controller action to replace the payum after capture one

Prometee committed:
Change tokenValue to payment request hash

Prometee committed:
Introduce PaymentRequest after pay

Prometee committed:
Rename service

Prometee committed:
Missing refactoring

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

Prometee committed:
Replace file header

Prometee committed:
Remove function import

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

Prometee committed:
Remove global function use

Prometee committed:
Refactor via quick syntax

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

Prometee committed:
Line return

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

Prometee committed:
Avoid publicly displaying available actions

Prometee committed:
Refactor to abstract class to avoid duplicating code

Prometee committed:
Fix indexed by action not type

Prometee committed:
Introduce parameterized url for the thank you page url

Prometee committed:
Move ShopBundle order_pay classes to CoreBundle

Prometee committed:
set authorize action when using use_authorize => true

Prometee committed:
Add some specs tests

Prometee committed:
Allow empty params and context

Prometee committed:
Fix public exposition

Prometee committed:
Fix wrong parameter name

Prometee committed:
Fix tag names on services

Prometee committed:
Fix empty locator service list throwing exception on the first ->get call

Prometee committed:
Add missing afterPayUrl provider and the related ShopBundle config param

Prometee committed:
Fix full getter

Prometee committed:
Use repository instead of manager

Prometee committed:
Fix wrong returned indexes

Prometee committed:
Add alias for the default provider

Prometee committed:
Allow no status command provider

Wojdylak committed:
Resolve conflicts between 2.0 and payment-request

Wojdylak committed:
[Api] Move PaymentRequestUpdateDataProviderSpec.php to disabledSpec directory

Wojdylak committed:
[Api] Remove unnecessary lines

Wojdylak committed:
[Api] Minor improvements

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#16858)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#16877)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

Prometee committed:
Missing final keyword on spec class

Prometee committed:
Fix phpstan

Wojdylak committed:
Resolve conflicts between 2.0 and payment-request

Wojdylak committed:
Minor improvement

Wojdylak committed:
Minor improvement

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#16889)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

Prometee committed:
Change route to put optional param at the end of the route

Prometee committed:
First check for Payum then check for PaymentRequest

Prometee committed:
Switch to the new MetadataInterface namespace

GSadee committed:
Merge branch '2.0' into payment-request

* 2.0: (21 commits)
  [CS][DX] Refactor
  [UserBundle] Add Jms/Serializer-bundle
  [UserBundle] FOS/Rest
  [CS][DX] Refactor
  [CS][DX] Refactor
  [Behat] Disable some customer scenarios
  [API] Split state provider configurations into admin and shop sections
  [Core] Add new payment repository method
  [Behat] Change the term status to state in payment and shipment scenarios
  [Behat] Refine codebase for shipment and payment contexts
  [Unit] Add new shipment and payment tests and refine existing ones
  [Behat] Adjust shipment and payment URLs
  [API][Shop] Secure shipment and payment endpoints
  [API][Shop] Move shipment and payment endpoints under the order resource
  [Unit] Add disableShopContext util function
  [Behat] Add requestGet method to API client
  [Unit] Change nAWw2jewpA token value to token
  [Unit] Add new response util functions
  [Unit] Refine order placer trait
  [Behat][API] Enable customer suite
  ...

GSadee committed:
[Behat] Fix typo in scenarios tags

GSadee committed:
Merge branch '2.0' into payment-request

* 2.0:
  [Docs] Fix link in grids customization guide

Wojdylak committed:
[API] Implement payment method group generator

GSadee committed:
Merge branch '2.0' into payment-request

* 2.0:
  ShopUserBasedExtension - rename function name
  [Behat] Enable shop checkout context in address book suite
  [Behat] Enable customer address book scenarios
  Refactor a bit the validation of Statistics, enable unit test
  [API] Cover missing suite for getting errors in admin's language
  [Api] Fix spec tests
  [Api] Enable behat tests
  [Api] Add change password operation
  [Api] Secure the customer get endpoint

TheMilek committed:
Upgrade PaymentRequests to APIP3

TheMilek committed:
Enable contract tests

TheMilek committed:
Replace data transformer with context builder and upgrade item provider to APIP3

TheMilek committed:
Update baseline and fix PHPStan

TheMilek committed:
Enable behat tests

TheMilek committed:
Remove getters from AddPaymentRequest

Prometee committed:
Refactor to externalize API PaymentRequest from PayPalContext

Prometee committed:
Introduce default action & payload providers when no action or payload are given

Prometee committed:
re-Enable some scenarios

Prometee committed:
Fix too many required arguments

Prometee committed:
Fix PHPUnit env test_cached_payum

Prometee committed:
Prioritize PaymentRequest

Prometee committed:
Simplify offline to the strict minimum

TheMilek committed:
Make the hash in UpdatePaymentRequest required

Prometee committed:
Rename var

Prometee committed:
Lets Payum first check if it's a Payum gateway for UI

Prometee committed:
One line import

Co-authored-by: Karol <33687392+Wojdylak@users.noreply.github.com>

GSadee committed:
Payment request UI (#16790)

| Q               | A
|-----------------|-----
| Branch?         | 2.0
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | yes
| Deprecations?   | yes
| Related tickets | none
| License         | MIT

This PR will allow `PaymentRequest` to be used through the UI (not the
API), the goal is to allow developper to create HTTP response they need
to handle any Payment Method. SyliusPayumBundle controller actions will
be preserved to avoid any changes for existing Payum gateways.

# Basic payment flow

```mermaid
---
title: Customer paying an order
---
flowchart TD
    A[sylius_shop_checkout_complete]
    B[sylius_shop_order_pay]
    subgraph Payment Provider
    C[External portal
    or
    custom page]
    end
    D[sylius_shop_order_after_pay]
    E[sylius_shop_order_thank_you]
    F[sylius_shop_order_show]
    A --> B
    B --> C
    C --> D
    D -- OK --> E
    D -- KO --> F
    F -- Choose another 
    payment method or retry --> B
```

# What is the current (Sylius 1.*) URLs flow

```mermaid
---
title: Sylius 1.* - Customer paying an order
---
flowchart TD
    A[sylius_shop_checkout_complete] --> B[sylius_shop_order_pay]
    subgraph Payum
    C[payum_capture_do]
    D{External URL
    or
    custom page}
    end
    E[sylius_shop_order_after_pay]
    subgraph Payum after capture
    F[Payum Status
    &
    ResolveNextRoute
    ]
    end
    G[sylius_shop_order_thank_you]
    H[sylius_shop_order_show]
    B -- No payment --> G
    B --> C
    C <--> D
    D --> E
    C --> E
    E --> F
    F -- KO --> H
    F -- OK --> G
    H -- Choose another 
    payment method or retry --> B
```

# What this PR is aiming to do

```mermaid
---
title: Sylius 2.* - Customer paying an order
---
flowchart TD
    A[sylius_shop_checkout_complete]
    B[sylius_shop_order_pay]
    subgraph No payment
    C[No payment processing]
    end
    subgraph Payum
    D[Keep existing Payum
    beforeCapture processing]
    end
    subgraph Payment Request
    E[Payment Request processing]
    F[Http Response processing]
    end
    G[sylius_shop_order_after_pay]
    subgraph Payum after pay
    H[Keep existing
    afterCapture processing]
    end
    subgraph Payment request after pay
    I[Payment request processing]
    J[Http response processing]
    end
    A --> B
    B --> C & D & E
    G --> H & I
    C --> K
    D --> G
    E --> F
    F --> G
    I --> J
    H -- OK --> K
    H -- KO --> L
    J -- OK --> K
    J -- KO --> L
    L -- Choose another 
    payment method or retry --> B
    K[sylius_shop_order_thank_you]
    L[sylius_shop_order_show]
```


**Order pay phase**
- [x] `sylius.controller.payum::prepareCaptureAction` dispatch action
content into several services.
- [x] Create a more generic
`sylius.controller.shop.order_pay::payAction` to replace the previous
SyliusPayumBundle controller action.
- [x] Introduce `PaymentRequest` orderPay HTTP Response.

**After pay phase**
- [x] `sylius.controller.payum::afterCaptureAction` dispatch action
content into several services.
- [x] Create a more generic
`sylius.controller.shop.order_pay::afterPayAction` to replace the
previous SyliusPayumBundle controller action.
- [x] Introduce `PaymentRequest` afterPay HTTP Response.

**Global**
- [ ] Creeate or update tests.

Wojdylak committed:
Resolve conflicts between 2.0 and payment-request

Wojdylak committed:
Remove paypal integration

Wojdylak committed:
Remove stripe integration

Wojdylak committed:
[UPMERGE] 2.0 -> payment-request (#16931)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

GSadee committed:
Resolve conflicts between 2.0 and payment-request

GSadee committed:
Remove usage of deprecated MessageHandlerInterface

mpysiak committed:
Add command bus for payment request

GSadee committed:
[CI] Fix matrix

GSadee committed:
[Payum] Remove empty test suite config

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17005)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17016)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

Wojdylak committed:
[Admin] Add payment request index and show pages

Wojdylak committed:
[Admin][PaymentRequest] Add behat tests

Wojdylak committed:
Mirror improvement

Wojdylak committed:
Fix AddPaymentRequest

GSadee committed:
Add command bus for payment request (#17014)

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | n/a
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
- Features, removing deprecations and BC breaks must be submitted
against the 2.0 branch
 - Make sure that the correct base branch is set

To be sure you are not breaking any Backward Compatibilities, check the
documentation:

https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

Wojdylak committed:
Move payment request routes to separate file

Wojdylak committed:
[PaymentRequest] Use constants in grid

Wojdylak committed:
[PaymentBundle] Add symfony/doctrine-bridge

Wojdylak committed:
[PaymentBundle] Move payment request to outside of payment

GSadee committed:
Resolve conflicts between 2.0 and payment-request

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17024)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

GSadee committed:
[API] Remove legacy ApiResources

GSadee committed:
[PaymentRequest] Move PaymentRequestAnnouncer to PaymentBundle

GSadee committed:
[PaymentRequest] Move PaymentRequestDuplicationChecker to PaymentBundle

Rafikooo committed:
[API] Remove legacy ApiResources (#17033)

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations? | no<!-- don't forget to update the UPGRADE-*.md file
-->
| Related tickets | 
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
- Features, removing deprecations and BC breaks must be submitted
against the 2.0 branch
 - Make sure that the correct base branch is set

To be sure you are not breaking any Backward Compatibilities, check the
documentation:

https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

GSadee committed:
[PaymentRequest] Move Payum related services to PayumBundle

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17041)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

GSadee committed:
[PaymentRequest] Move Payum processors to PayumBundle

GSadee committed:
[PaymentRequest] Move PaymentRequestHashAware interface and trait to PaymentBundle

GSadee committed:
[PaymentRequest] Move Payum commands and command handlers to PayumBundle

Wojdylak committed:
[PaymentBundle] Minor improvements

Wojdylak committed:
[PaymentBundle] Minor improvement

GSadee committed:
[PaymentRequest] Move Payum command providers to PayumBundle

GSadee committed:
[PaymentRequest] Move command providers to PaymentBundle

GSadee committed:
[PaymentRequest] Move offline command and command handlers to PaymentBundle

GSadee committed:
[PaymentRequest] Move providers to PaymentBundle

GSadee committed:
[PaymentRequest] Adjust interface in messenger routing configuration

GSadee committed:
[PaymentRequest] Remove unneeded payment_request services file in CoreBundle

GSadee committed:
[Admin] Add payment request index and show pages (#17020)

| Q               | A
|-----------------|-----
| Branch?         | 2.0
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | n/a
| License         | MIT


![image](https://github.com/user-attachments/assets/f58d4c3c-2543-4fd2-860a-a5025dc318b9)

![image](https://github.com/user-attachments/assets/75566240-5bc7-4041-a3aa-4a2335478694)

GSadee committed:
[PaymentRequest] Adjust class namespace in test_cached_payum config

GSadee committed:
[Composer] Revert changing PayumBundle description

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17050)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

GSadee committed:
[Composer] Move messenger configuration for payment requests to PaymentBundle

GSadee committed:
Fix test service definition for test_cached_payum env

GSadee committed:
[PaymentRequest] Move repository configuration to PaymentBundle

GSadee committed:
Fix test application for PaymentBundle

GSadee committed:
[PaymentRequest] Apply naming suggestions after code review

Wojdylak committed:
[PaymentBundle] Add state machine for paymentRequest

Wojdylak committed:
Change default adapter to symfony_workflow for sylius_state_machine_abstraction

Wojdylak committed:
[SyliusStateMachine] Fix configuration test

Wojdylak committed:
[PaymentBundle] Add sylius-state-machine-abstraction

GSadee committed:
Merge branch 'payment-request' into upmerge/2.0_payment-request

GSadee committed:
[PaymentRequest] Apply naming suggestions after code review in PaymentBundle

GSadee committed:
Remove fos_rest config file form test_cached_payum env

mpysiak committed:
[PaymentRequest] Split the implementation and move it to appropriate bundles (#17037)

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations? | no<!-- don't forget to update the UPGRADE-*.md file
-->
| Related tickets | replaces #15978
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
- Features, removing deprecations and BC breaks must be submitted
against the 2.0 branch
 - Make sure that the correct base branch is set

To be sure you are not breaking any Backward Compatibilities, check the
documentation:

https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

Wojdylak committed:
[PaymentBundle] Add winzou state machine for paymentRequest

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17065)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

Wojdylak committed:
Move payment request state machine to CoreBundle

Wojdylak committed:
Use CoreBundle instead of PaymentBundle

GSadee committed:
[PaymentBundle] Add state machine for paymentRequest (#17068)

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | n/a
| License         | MIT

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17080)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

GSadee committed:
[Composer] Remove dependency to payum/core in Core component

GSadee committed:
Move Payum-related exception to PayumBundle

GSadee committed:
[OrderPay] Move Payum-related providers to PayumBundle

GSadee committed:
Move Payum-related translations to PayumBundle

GSadee committed:
[API] Remove non existing exception from status mapping

GSadee committed:
[API] Fix GatewayConfig class in serialization

GSadee committed:
Decouple CoreBundle from PayumBundle

GSadee committed:
[Composer] Change dependency from Core component to CoreBundle in PayumBundle

TheMilek committed:
Add listener for every change of payment method

TheMilek committed:
Cancel every payment request that has old payment method and is not completed or failed

TheMilek committed:
Add behat test

TheMilek committed:
minor ecs fixes

GSadee committed:
[OrderPay] Refactor services files in ShopBundle

GSadee committed:
Bring back SyliusPayumBundle to Admin and Api bundles

GSadee committed:
Move Payum-related configs and routings from ShopBundle to PayumBundle

TheMilek committed:
Update PaymentRequestCanceller

TheMilek committed:
Move the Listener from API to PaymentBundle to make it work globally on all payments

TheMilek committed:
Add behat tests for UI

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17090)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

NoResponseMate committed:
minor #17083 Decouple CoreBundle from PayumBundle + related improvements (GSadee)

This PR was merged into the payment-request branch.

Discussion
----------

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | yes
| Deprecations?   | no<!-- don't forget to update the UPGRADE-*.md file -->
| Related tickets | 
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
 - Features, removing deprecations and BC breaks must be submitted against the 2.0 branch
 - Make sure that the correct base branch is set

 To be sure you are not breaking any Backward Compatibilities, check the documentation:
 https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->


Commits
-------
  [Composer] Remove dependency to payum/core in Core component
  Move Payum-related exception to PayumBundle
  [OrderPay] Move Payum-related providers to PayumBundle
  Move Payum-related translations to PayumBundle
  [API] Remove non existing exception from status mapping
  [API] Fix GatewayConfig class in serialization
  Decouple CoreBundle from PayumBundle
  [Composer] Change dependency from Core component to CoreBundle in PayumBundle
  [OrderPay] Refactor services files in ShopBundle
  Bring back SyliusPayumBundle to Admin and Api bundles
  Move Payum-related configs and routings from ShopBundle to PayumBundle

TheMilek committed:
Add parameter for payment request states that should be cancelled

GSadee committed:
[PaymentRequest] Automatically cancel all pending payment requests when the payment method gets changed (#15950)

| Q | A |

|-----------------|--------------------------------------------------------------|
| Branch? | payment-request <!-- see the comment below --> |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| License | MIT |

<!--
 - Bug fixes must be submitted against the 1.12 branch
 - Features and deprecations must be submitted against the 1.13 branch
 - Make sure that the correct base branch is set

To be sure you are not breaking any Backward Compatibilities, check the
documentation:

https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

GSadee committed:
Resolve conflicts between 2.0 and payment-request

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17113)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

TheMilek committed:
Correct the parameter name for states that should be cancelled, when payment method is changed

Prometee committed:
Remove remaining empty file

GSadee committed:
Correct the parameter name for states that should be cancelled, when payment method is changed (#17130)

| Q               | A
|-----------------|-----
| Branch?         | payment-request <!-- see the comment below -->
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
- Features, removing deprecations and BC breaks must be submitted
against the 2.0 branch
 - Make sure that the correct base branch is set

To be sure you are not breaking any Backward Compatibilities, check the
documentation:

https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->
continuation of:

https://github.com/Sylius/Sylius/pull/15950

GSadee committed:
[Core] Bring back importing PaymentBundle config

GSadee committed:
Resolve conflicts between 2.0 and payment-request

GSadee committed:
[Payment] Bring back winzouStateMachineBundle to test application's bundles

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17134)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

Prometee committed:
Add 2 webhooks routes

Prometee committed:
Remove shop prefix

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

Prometee committed:
Remove use and add file header

Prometee committed:
Remove composite suffix

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

Prometee committed:
Global renaming to Notify

# Conflicts:
#	src/Sylius/Bundle/PaymentBundle/Resources/config/services/controllers.xml

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17154)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

GSadee committed:
Resolve conflicts between 2.0 and payment-request

GSadee committed:
[Admin] Fix Twig Hooks config

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17156)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

Prometee committed:
Add Attribute for autoconfiguration

Prometee committed:
Add plural to the route

GSadee committed:
Merge branch '2.0' into payment-request

* 2.0: (22 commits)
  Add the packages that have become optional in version 1.12 of the ResourceBundle.
  Format tags in services
  Fixes after CR
  Add image lightbox
  [Shop][UI] Ensure correct button disabling based on available shipping methods
  [Behat] Implement refined prevent shipping step completion scenarios
  [Behat] Refine Prevent shipping step completion scenarios
  [Behat] Add patch method to ApiClientInterface
  [Behat] Update preventing payment step feature with revised scenarios and implementation
  [Behat] Implement prevent shipping step completion scenarios
  [Behat] Add AddressFactory for default address setup
  [Behat] Update preventing shipping step feature with revised scenarios and improved phrasing
  [Behat][API] Enable preventing shipping step completion scenarios
  Fix unit tests
  Fix reviews test
  Add trait to live components
  Fix liveCollection missing template
  [Admin] Move templates to twig_hook configuration
  [UI] Fix phpunit tests
  [Shop] Move template configuration to twig_hooks
  ...

GSadee committed:
Merge branch '2.0' into payment-request

* 2.0:
  Update README.md

GSadee committed:
[CI] Use default env in MySQL workflow

Prometee committed:
Add a new service to wrap the non serializable Symfony request

Prometee committed:
Use symfony request wrapper

Prometee committed:
Do not re-run the same notify when it was already completed

Prometee committed:
Fix routes

Prometee committed:
Update service definition

Prometee committed:
Rename property

Prometee committed:
Add Dummy services to handle notify on offline payment method

Prometee committed:
Fix wrong order number generation

Prometee committed:
New context to handle Payment request notify urls

Prometee committed:
CS fix

Prometee committed:
Fix ecs

Prometee committed:
Array definition

Prometee committed:
Remove null type

Prometee committed:
Simplify if statement

Prometee committed:
Avoid using specific final state, instead check if it is one

Prometee committed:
Rename wrapper to normalizer

Prometee committed:
Move final transition checker to a dedicated service and add it to the announcer

Prometee committed:
Introduce a new service to handle request payload processing

Prometee committed:
Ease the autowiring

Prometee committed:
Missing file header

Prometee committed:
Add missing NotifyPaymentRequest handler for Payum

Prometee committed:
Introduce a new service to handle Notify response

GSadee committed:
Merge branch '2.0' into payment-request

* 2.0:
  [CS][DX] Refactor
  Fix session unavailable in locale context
  [API] Change argument from string to array of prefixes in documentation modifier for removing routes
  [API] Add documentation modifier for removing non-API routes

mpysiak committed:
[CI] Use default env in MySQL workflow (#17172)

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations? | no<!-- don't forget to update the UPGRADE-*.md file
-->
| Related tickets | 
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
- Features, removing deprecations and BC breaks must be submitted
against the 2.0 branch
 - Make sure that the correct base branch is set

To be sure you are not breaking any Backward Compatibilities, check the
documentation:

https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

GSadee committed:
Revert "[CI] Use default env in MySQL workflow"

GSadee committed:
Revert "[CI] Use default env in MySQL workflow" (#17176)

Reverts Sylius/Sylius#17172

GSadee committed:
Merge branch '2.0' into payment-request

* 2.0:
  [UI] Add default component template config tests

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17183)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

GSadee committed:
[API] Fix configuration of Payment Request resource

Prometee committed:
Rename RequestPayload processor to NotifyPayload and use array_merge

Prometee committed:
Add test for NotifyPayloadProcessor

GSadee committed:
[API] Refactor Payment Request POST endpoint

GSadee committed:
[API] Add unit test for AddPaymentRequestHandler

Prometee committed:
Remove docblock

GSadee committed:
[PaymentRequest] Add experimental tag to the classes in Payment bundle and component

GSadee committed:
[PaymentRequest] Add experimental tag to the classes in PayumBundle

Prometee committed:
Rename PaymentNotify to NotifyPayment

GSadee committed:
[OrderPay] Add experimental tag to the classes in CoreBundle

Prometee committed:
Fix ECS

GSadee committed:
[PaymentRequest] Add experimental tag to the classes in ApiBundle

NoResponseMate committed:
minor #17198 [API] Fix configuration of Payment Request resource + refactor POST endpoint (GSadee)

This PR was merged into the payment-request branch.

Discussion
----------

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no<!-- don't forget to update the UPGRADE-*.md file -->
| Related tickets | 
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
 - Features, removing deprecations and BC breaks must be submitted against the 2.0 branch
 - Make sure that the correct base branch is set

 To be sure you are not breaking any Backward Compatibilities, check the documentation:
 https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->


Commits
-------
  [API] Fix configuration of Payment Request resource
  [API] Refactor Payment Request POST endpoint
  [API] Add unit test for AddPaymentRequestHandler

Prometee committed:
Rename PaymentRequestFinalTransitionChecker to FinalizedPaymentRequestChecker

GSadee committed:
[Payment-Request] Notify routes (#17151)

| Q               | A
|-----------------|-----
| Branch?         | 2.0
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | none
| License         | MIT

This PR add two new routes :

```yaml
# Classic notify
sylius_payment_method_notify:
    path: /payment-methods/{code}

# Notify using an existing PaymentRequest
sylius_payment_request_notify:
    path: /payment-requests/{hash}
```

GSadee committed:
[UPMERGE] 2.0 -> payment-request (#17207)

This PR has been generated automatically.
For more details see
[upmerge_pr.yaml](/Sylius/Sylius/blob/1.13/.github/workflows/upmerge_pr.yaml).

**Remember!** The upmerge should always be merged with using `Merge pull
request` button.

In case of conflicts, please resolve them manually with usign the
following commands:
```
git fetch upstream
gh pr checkout <this-pr-number>
git merge upstream/payment-request -m "Resolve conflicts between 2.0 and payment-request"
```

If you use other name for the upstream remote, please replace `upstream`
with the name of your remote pointing to the `Sylius/Sylius` repository.

Once the conflicts are resolved, please run `git merge --continue` and
push the changes to this PR.

NoResponseMate committed:
minor #17201 [PaymentRequest] Add experimental tag to the classes (GSadee)

This PR was merged into the payment-request branch.

Discussion
----------

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no<!-- don't forget to update the UPGRADE-*.md file -->
| Related tickets | 
| License         | MIT

As Payment Requests functionality is complex, we would like to gather feedback from real-world usage and adapt it in accordance with the feedback, so I'm marking this feature and its classes as `experimental`.
<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
 - Features, removing deprecations and BC breaks must be submitted against the 2.0 branch
 - Make sure that the correct base branch is set

 To be sure you are not breaking any Backward Compatibilities, check the documentation:
 https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->


Commits
-------
  [PaymentRequest] Add experimental tag to the classes in Payment bundle and component
  [PaymentRequest] Add experimental tag to the classes in PayumBundle
  [OrderPay] Add experimental tag to the classes in CoreBundle
  [PaymentRequest] Add experimental tag to the classes in ApiBundle

GSadee committed:
Merge branch '2.0' into payment-request

* 2.0:
  Change sort_by filter to sylius_sort_by
  Fix note in the UPGRADE file
  [Composer] Bump TwigExtra to the 0.4 version
  [UiBundle] Deprecate classes and services moved to TwigExtra package
  Update upgrade file
  [UserBundle] Unification of service names
  Update upgrade file
  [TaxonomyBundle] Unification os service names
  Add note to UPGRADE guide
  Use Twig Extra package
  Update upgrade file
  [ShippingBundle] Unification of service name
  Update upgrade file
  [ReviewBundle] Unification of service names
  [PromotionBundle] Unification of sylius.registry_promotion_action
  Update upgrade file
  [TaxationBundle] Unification of service names

NoResponseMate commented on UPGRADE-2.0.md:
```diff
@@ -6,6 +6,9 @@
     * The `sylius_default.bus` and `sylius_event.bus` configuration options were removed. Use `sylius.command_bus` and
       `sylius.event_bus` for commands and events respectively.
 
+* SyliusStateMachineAbstraction:
+    * The `sylius_state_machine_abstraction.default_adapter` option has been changed from `winzou_state_machine` to `symfony_workflow`.
+
```
This doesn't seem to be related

GSadee replied:
Directly no, but it was added on this branch, so there is no point in reversing it 🖖🏻

NoResponseMate commented on config/routes/test_cached_payum/routing.yaml:
```diff
@@ -0,0 +1,5 @@
+sylius_test_plugin_main:
+    path: /test/main
+    controller: Symfony\Bundle\FrameworkBundle\Controller\TemplateController::templateAction
+    defaults:
+        template: "@SyliusTestPlugin/main.html.twig"
```
Is this needed?

GSadee replied:
I guess so, it is added also in other test envs and if I see correctly, it is used for some Behat scenarios for theming 🤔

NoResponseMate commented on config/services_test_cached_payum.yaml:
```diff
@@ -0,0 +1,14 @@
+    ##
+    # Prioritize Payum by tagging services with a higher priority (or lower for tagged_iterator).
+    ##
+    
```
```suggestion
    # Prioritize Payum by tagging services with a higher priority (or lower for tagged_iterator).
```

NoResponseMate commented on features/admin/product/managing_products/sorting_products_within_a_taxon_by_position.feature:
```diff
@@ -80,7 +80,7 @@ Feature: Sorting listed products from a taxon by position
         And this product should be at position 16
 
-    @no-api @ui @mink:chromedriver
+    @no-api @tod-ui @mink:chromedriver
```
What happened here?

NoResponseMate commented on phpstan-baseline.neon:
```diff
@@ -1100,6 +1100,26 @@ parameters:
 			path: src/Sylius/Bundle/CoreBundle/Order/Checker/OrderPromotionsIntegrityChecker.php
 
+		-
+			message: "#^Call to method toBinary\\(\\) on an unknown class Symfony\\\\Component\\\\Uid\\\\Uuid\\.$#"
```
This seems wrong, but we'll have to regenerate the baseline either way so 🤷

NoResponseMate commented on src/Sylius/Behat/Context/Api/Shop/PaymentRequestContext.php:
```diff
@@ -0,0 +1,122 @@
+        $response = $this->client->getLastResponse();
+
+        Assert::same($this->responseChecker->getValue($response, 'action'), $action, sprintf('Payment request should have action %s', $action));
+        Assert::true(str_contains($this->responseChecker->getValue($response, 'method'), $paymentMethod->getCode()), sprintf('Payment request should have payment method %s', $paymentMethod->getCode()));
```
```suggestion
        Assert::contains($this->responseChecker->getValue($response, 'method'), $paymentMethod->getCode(), sprintf('Payment request should have payment method %s', $paymentMethod->getCode()));
```

NoResponseMate commented on src/Sylius/Behat/Resources/config/services/contexts/api/shop.xml:
```diff
@@ -161,6 +161,16 @@
+            <argument type="service" id="sylius.behat.request_factory" />
+            <argument type="service" id="sylius.repository.payment_request" />
+            <argument type="service" id="sylius.behat.page.shop.payment_request.payment_method_notify" />
+            <argument type="service" id="test.client" />
```
```suggestion
```

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/templates/order/show/content/sections/payments/item/actions/list_payment_requests.html.twig:
```diff
@@ -0,0 +1,7 @@
+
+{% set payment = hookable_metadata.context.payment %}
+
+<a title="details" href="{{ path('sylius_admin_payment_request_index', {'paymentId': payment.id}) }}" class="btn btn-icon" data-bs-toggle="tooltip" data-bs-title="{{ 'sylius.ui.list_payment_requests'|trans }}" {{ sylius_test_html_attribute('list_payment_requests') }}>
```
```suggestion
<a title="details" href="{{ path('sylius_admin_payment_request_index', {'paymentId': payment.id}) }}" class="btn btn-icon" data-bs-toggle="tooltip" data-bs-title="{{ 'sylius.ui.list_payment_requests'|trans }}" {{ sylius_test_html_attribute('list-payment-requests') }}>
```

NoResponseMate commented on src/Sylius/Behat/Page/Admin/Payment/IndexPage.php:
```diff
@@ -48,6 +48,17 @@ public function showOrderPageForNthPayment(int $position): void
+        $row = $tableAccessor->getRowsWithFields($table, [])[$position];
+        $field = $tableAccessor->getFieldFromRow($table, $row, 'actions');
+
+        $field->find('css', '[data-test-show-action="List payment requests"]')->click();
```
```suggestion
        $field->find('css', '[data-test-list-payment-requests]')->click();
```

GSadee replied:
I have left it as it is, because here the generic show action on the grid is used

NoResponseMate commented on src/Sylius/Bundle/ApiBundle/Command/Payment/AddPaymentRequest.php:
```diff
@@ -0,0 +1,31 @@
+{
+    public function __construct(
+        public readonly string $orderTokenValue,
+        public readonly int|string $paymentId,
```
```suggestion
        public readonly mixed $paymentId,
```
?

NoResponseMate commented on src/Sylius/Bundle/ApiBundle/EventSubscriber/PaymentRequestEventSubscriber.php:
```diff
@@ -0,0 +1,52 @@
+
+        if (!$paymentRequest instanceof PaymentRequestInterface) {
+            return;
+        }
```
```suggestion
        $paymentRequest = $event->getControllerResult();
        if (!$paymentRequest instanceof PaymentRequestInterface) {
            return;
        }
```

NoResponseMate commented on src/Sylius/Bundle/ApiBundle/EventSubscriber/PaymentRequestEventSubscriber.php:
```diff
@@ -0,0 +1,52 @@
+
+        $method = $event->getRequest()->getMethod();
+
+        if (in_array($method, [Request::METHOD_POST, Request::METHOD_PUT, Request::METHOD_PATCH], true)) {
```
```suggestion
        if (in_array($event->getRequest()->getMethod(), [Request::METHOD_POST, Request::METHOD_PUT, Request::METHOD_PATCH], true)) {
```

NoResponseMate commented on src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PaymentRequest.xml:
```diff
@@ -0,0 +1,56 @@
+                uriTemplate="/admin/payment-requests/{hash}"
+            >
+                <normalizationContext>
+                    <values>
+                        <value>sylius:admin:payment_request:show</value>
+                    </values>
```
Seems it's also missing the `groups` node 💃

NoResponseMate commented on src/Sylius/Bundle/ApiBundle/Validator/Constraints/ChosenPaymentRequestActionEligibilityValidator.php:
```diff
@@ -0,0 +1,85 @@
+    private function getCommandProvider(AddPaymentRequest $addPaymentRequest): ?PaymentRequestCommandProviderInterface
+    {
+        /** @var PaymentMethodInterface|null $paymentMethod */
+        $paymentMethod = $this->paymentMethodRepository->findOneBy(['code' => $addPaymentRequest->paymentMethodCode]);
+
+        if ($paymentMethod === null) {
+            return null;
+        }
+
+        $gatewayConfig = $paymentMethod->getGatewayConfig();
+        if (null === $gatewayConfig) {
+            return null;
+        }
```
```suggestion
        $paymentMethod = $this->paymentMethodRepository->findOneBy(['code' => $addPaymentRequest->paymentMethodCode]);
        if ($paymentMethod?->getGatewayConfig() === null) {
            return null;
        }
```

NoResponseMate commented on src/Sylius/Bundle/ApiBundle/Validator/Constraints/ChosenPaymentRequestActionEligibilityValidator.php:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        $factoryName = $this->gatewayFactoryNameProvider->provide($paymentMethod);
+        $commandProvider = $this->gatewayFactoryCommandProvider->getCommandProvider($factoryName);
+
+        return $commandProvider ?? null;
```
```suggestion

        return $this->gatewayFactoryCommandProvider->getCommandProvider($factoryName);
```

NoResponseMate commented on src/Sylius/Bundle/ApiBundle/Resources/config/serialization/PaymentRequest.xml:
```diff
@@ -0,0 +1,71 @@
+            xsi:schemaLocation="http://symfony.com/schema/dic/serializer-mapping https://symfony.com/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd"
+>
+    <class name="Sylius\Component\Payment\Model\PaymentRequest">
+        <attribute name="id" ignore="true" />
+
```
```suggestion
```

GSadee replied:
It seems to be needed, otherwise we get:
```
request.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "Invalid identifier value or configuration." at ReadListener.php line 89 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): Invalid identifier value or configuration. at /Users/gsadee/Sites/Sylius/Sylius/vendor/api-platform/core/src/Symfony/EventListener/ReadListener.php:89)\n[previous exception] [object] (ApiPlatform\\Metadata\\Exception\\InvalidIdentifierException(code: 0): Parameter \"id\" not found, check the identifiers configuration. at /Users/gsadee/Sites/Sylius/Sylius/vendor/api-platform/core/src/State/UriVariablesResolverTrait.php:41)"} []
```

NoResponseMate commented on src/Sylius/Bundle/ApiBundle/Checker/UpdatePaymentRequestEligibilityChecker.php:
```diff
@@ -0,0 +1,24 @@
+
+use Sylius\Component\Payment\Model\PaymentRequestInterface;
+
+final class UpdatePaymentRequestEligibilityChecker implements UpdatePaymentRequestEligibilityCheckerInterface
+{
+    public function isEligible(PaymentRequestInterface $paymentRequest): bool
+    {
+        return in_array($paymentRequest->getState(), [PaymentRequestInterface::STATE_NEW, PaymentRequestInterface::STATE_PROCESSING]);
+    }
+}
```
Guess we could think about using the newly introduced `FinalizedPaymentRequestChecker` in place of this one, as it's pretty much the same logic, only the other way around. Less code necessary to customize for the end users as well.

NoResponseMate commented on src/Sylius/Bundle/CoreBundle/OrderPay/Provider/NoPaymentPayResponseProvider.php:
```diff
@@ -0,0 +1,48 @@
+        RequestConfiguration $requestConfiguration,
+        OrderInterface $order
+    ): Response {
+        $url = $this->orderPayFinalUrlProvider->getUrl(null);
+
+        return new RedirectResponse($url);
```
```suggestion
        return new RedirectResponse($this->orderPayFinalUrlProvider->getUrl(null));
```

NoResponseMate commented on src/Sylius/Bundle/CoreBundle/OrderPay/Provider/NoPaymentPayResponseProvider.php:
```diff
@@ -0,0 +1,48 @@
+        RequestConfiguration $requestConfiguration,
+        OrderInterface $order
+    ): bool {
+        $payment = $this->paymentToPayResolver->getLastPayment($order);
+
+        return null === $payment;
```
```suggestion
        return null === $this->paymentToPayResolver->getLastPayment($order);
```

NoResponseMate commented on src/Sylius/Bundle/CoreBundle/OrderPay/Provider/Offline/StatusHttpResponseProvider.php:
```diff
@@ -0,0 +1,46 @@
+    ): bool {
+        return $paymentRequest->getAction() === PaymentRequestInterface::ACTION_STATUS;
+    }
+
+    public function getResponse(
+        RequestConfiguration $requestConfiguration,
+        PaymentRequestInterface $paymentRequest,
+    ): Response {
+        // Force null payment to go to the thank you page
+        $finalUrl = $this->finalUrlProvider->getUrl(null);
+
+        return new RedirectResponse($finalUrl);
+    }
```
```suggestion

    // Forces null payments to go to the thank you page
    public function getResponse(
        RequestConfiguration $requestConfiguration,
        PaymentRequestInterface $paymentRequest,
    ): Response {
        return new RedirectResponse($this->finalUrlProvider->getUrl(null));
    }
```

NoResponseMate commented on src/Sylius/Bundle/CoreBundle/OrderPay/Resolver/PaymentToPayResolverInterface.php:
```diff
@@ -0,0 +1,23 @@
+/** @experimental */
+interface PaymentToPayResolverInterface
+{
+    public function getLastPayment(OrderInterface $order): ?PaymentInterface;
```
```suggestion
    public function getPayment(OrderInterface $order): ?PaymentInterface;
```

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Canceller/PaymentRequestCanceller.php:
```diff
@@ -0,0 +1,55 @@
+                );
+
+                $this->objectManager->persist($paymentRequest);
+                $this->objectManager->flush();
+            }
+        }
```
```suggestion
            }
        }

        $this->objectManager->flush();
```
Flushing in a foreach is a bad idea

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Checker/FinalizedPaymentRequestChecker.php:
```diff
@@ -0,0 +1,39 @@
+
+    public function isFinal(PaymentRequestInterface $paymentRequest): bool
+    {
+        $state = $paymentRequest->getState();
+        $nextTransition = $this->stateMachine->getTransitionFromState(
+            $paymentRequest,
+            PaymentRequestTransitions::GRAPH,
+            $state,
+        );
+
+        return null === $nextTransition;
```
```suggestion
        return null === $this->stateMachine->getTransitionFromState(
            $paymentRequest,
            PaymentRequestTransitions::GRAPH,
            $paymentRequest->getState(),
        );
```

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Doctrine/ORM/PaymentRequestRepository.php:
```diff
@@ -0,0 +1,95 @@
+            ->andWhere('o.action = :action')
+            ->andWhere('o.payment = :payment')
+            ->andWhere('o.method = :method')
+            // Symfony\Bridge\Doctrine\Types\UuidType has signature changes between sf 5.4 and 6.4
+            // We are forced to use type: "uuid" to use both Sf versions
+            ->setParameter('paymentRequest', $paymentRequest->getHash(), 'uuid')
```
5.4 went away, so I guess this can be resolved 🤔

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Form/Type/PaymentMethodType.php:
```diff
@@ -37,10 +55,40 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $resolver->setDefaults([
+            'validation_groups' => function (FormInterface $form): array {
+                $data = $form->getData();
+
+                return $this->paymentMethodGroupsGenerator->__invoke($data);
```
```suggestion
                return $this->paymentMethodGroupsGenerator->__invoke($form->getData());
```

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Generator/GatewayConfigNameGenerator.php:
```diff
@@ -0,0 +1,29 @@
+
+use Sylius\Component\Payment\Model\PaymentMethodInterface;
+
+final readonly class GatewayConfigNameGenerator implements GatewayConfigNameGeneratorInterface
```
```suggestion
final readonly class GatewayNameGenerator implements GatewayNameGeneratorInterface
```

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Provider/AbstractServiceProvider.php:
```diff
@@ -0,0 +1,73 @@
+use Symfony\Contracts\Service\ServiceProviderInterface;
+
+/** @experimental */
+abstract class AbstractServiceProvider implements ServiceProviderAwareProviderInterface
```
We need to think about simplifying these, and maybe split the logic into a few different services, as it is it's pretty magical and it's easy to get lost

GSadee replied:
We will have to rethink this service and its implementation and probably refactor it, but after the merge of this branch into 2.0 🖖🏻

Prometee commented on src/Sylius/Bundle/ApiBundle/Command/Payment/AddPaymentRequest.php:
```diff
@@ -0,0 +1,31 @@
+
+namespace Sylius\Bundle\ApiBundle\Command\Payment;
+
+use Sylius\Bundle\ApiBundle\Attribute\OrderTokenValueAware;
+use Sylius\Bundle\ApiBundle\Command\IriToIdentifierConversionAwareInterface;
+
+/** @experimental */
+#[OrderTokenValueAware]
+class AddPaymentRequest implements IriToIdentifierConversionAwareInterface
```
```suggestion
use Sylius\Bundle\ApiBundle\Attribute\OrderTokenValueAware;
use Sylius\Bundle\ApiBundle\Attribute\PaymentIdAware;
use Sylius\Bundle\ApiBundle\Command\IriToIdentifierConversionAwareInterface;

/** @experimental */
#[OrderTokenValueAware]
#[PaymentIdAware]
class AddPaymentRequest implements IriToIdentifierConversionAwareInterface
```

To complete the suggestion of @NoResponseMate just bellow.

GSadee replied:
Adding this makes no sense as it would not be used at all, as we do not have payment id set as a parameter in the URL 🖖🏻

GSadee committed:
Regenerate phpstan-baseline.neon

GSadee committed:
[API] Fix configuration of Payment Request resource

GSadee committed:
[PaymentRequest] Minor improvements

GSadee committed:
[PaymentRequest] Refactor AddPaymentRequest command

GSadee committed:
[Payment] Rename GatewayConfigNameGenerator to GatewayNameGenerator

GSadee committed:
[API] Remove UpdatePaymentRequestEligibilityChecker in favour of FinalizedPaymentRequestChecker

Wojdylak committed:
[PaymentRequest] Fixes and improvements before merging to 2.0 (#17217)

| Q               | A
|-----------------|-----
| Branch?         | payment-request
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations? | no<!-- don't forget to update the UPGRADE-*.md file
-->
| Related tickets | fixes suggestions from #17165 
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.13 branch
 - Features and deprecations must be submitted against the 1.14 branch
- Features, removing deprecations and BC breaks must be submitted
against the 2.0 branch
 - Make sure that the correct base branch is set

To be sure you are not breaking any Backward Compatibilities, check the
documentation:

https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

GSadee committed:
Merge branch '2.0' into payment-request

* 2.0:
  [CS][DX] Refactor
  [Shop] Rename sylius_logo.html.twig to logo.html.twig

Rafikooo commented on features/shop/payment_request/calling_notify_for_a_payment_request.feature:
```diff
@@ -0,0 +1,29 @@
+        And there is a customer "john@example.com" that placed order with "PHP T-Shirt" product to "United States" based billing address with "Free" shipping method and "Offline" payment method
+
+
+    @ui
```
The @api / @no-api tag should be added

GSadee replied:
https://github.com/Sylius/Sylius/pull/17151#discussion_r1796888567
